### PR TITLE
Fixed #26060 -- Fixed regression for OneToOneField backward reference in a read-only field in the admin

### PR DESCRIPTION
**Pull request by sasha0**

### CODE REVIEW

File: tests/admin_views/tests.py [hunk 1]
```diff
@@ -4706,6 +4707,14 @@ def test_readonly_backwards_ref(self):
         response = self.client.get(reverse('admin:admin_views_topping_add'))
         self.assertEqual(response.status_code, 200)
 
+    @override_settings(DEBUG_PROPAGATE_EXCEPTIONS=True)
+    def test_readonly_onetoone_backwards_ref(self):
+        """Regression test for #26060"""
```
Comment (timgraham) [inline tests/admin_views/tests.py]: My preference is to describe the expected behavior in the docstring and include a ticket reference only if it's an obscure issue that benefits from the additional context of the ticket.

Comment (sasha0) [inline tests/admin_views/tests.py]: >My preference is to describe the expected behavior in the docstring and include a ticket reference only if it's an obscure issue that benefits from the additional context of the ticket.

Ok, will update.

Could you please suggest a potential reason why exception might be not propagating in the test? Even though it has necessary setting enabled and exception being raised for sure in `django.contrib.admin.helpers.AdminReadonlyField.contents` method?
Thanks.

Comment (timgraham) [inline tests/admin_views/tests.py]: Trying adding `@override_settings(DEBUG=True)` to the test. The template system might be silencing the exception otherwise and rendering an empty string otherwise.

Comment (sasha0) [inline tests/admin_views/tests.py]: That helped, thanks!

File: tests/admin_views/models.py [hunk 2]
```diff
@@ -335,6 +335,14 @@ class Vodcast(Media):
     released = models.BooleanField(default=False)
 
 
+class Screencast(Media):
+    media = models.OneToOneField(Media, models.CASCADE, primary_key=True, related_name='screencast')
+    title = models.CharField(max_length=100)
+
+    def __str__(self):
+        return self.title
```
Comment (charettes) [inline tests/admin_views/models.py]: You can reuse the `PlotDetails` model instead of defining a new one as it slow downs the test suite execution.

Comment (sasha0) [inline tests/admin_views/models.py]: Good idea, thanks.

File: tests/admin_views/tests.py [hunk 3]
```diff
@@ -4688,6 +4688,18 @@ def test_readonly_backwards_ref(self):
         response = self.client.get(reverse('admin:admin_views_topping_add'))
         self.assertEqual(response.status_code, 200)
 
+    @override_settings(DEBUG=True)
```
Comment (timgraham) [inline tests/admin_views/tests.py]: I guess `DEBUG=True` isn't needed at this point since `assertContains()` would fail without it, correct?

File: docs/releases/1.8.9.txt [hunk 4]
```diff
@@ -26,3 +26,5 @@ Bugfixes
 
 * Fixed a crash when using an ``__in`` lookup inside a ``Case`` expression
   (:ticket:`26071`).
+* Fixed a regression that caused exception for backward reference of a OneToOneField
```
Comment (timgraham) [inline docs/releases/1.8.9.txt]:
```
* Fixed a crash when using a reverse ``OneToOneField`` in
  ``ModelAdmin.readonly_fields`` (:ticket:`26060`).
```
Please add a blank line between the bullets too.

File: tests/admin_views/tests.py (hunk 3 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Right.
However, without it, exception won't raise and `self.assertEqual(response.status_code, 200)` would always pass.

File: django/contrib/admin/helpers.py [hunk 5]
```diff
@@ -214,6 +214,8 @@ def contents(self):
             else:
                 if isinstance(f.remote_field, ManyToManyRel) and value is not None:
                     result_repr = ", ".join(map(six.text_type, value.all()))
+                elif isinstance(f.remote_field, OneToOneField) and value is not None:
```
Comment (timgraham) [inline django/contrib/admin/helpers.py]: Should add a test to show why `value is not None` is necessary.

Comment (sasha0) [inline django/contrib/admin/helpers.py]: Separate one?

Comment (timgraham) [inline django/contrib/admin/helpers.py]: Whatever makes sense. I'm actually think it's not needed but it seems we need a test to verify what happens when the reverse relation is null.

**Comment by timgraham** [comment]
I don't think the current fix replicates the behavior in older versions of Django. It seems that reverse O2O would go to the branch with `result_repr = smart_text(value)` before. However, I'm not sure if there are any practical differences between `text_type()` and `smart_text()` in this case.

**Comment by sasha0** [comment]
Ok, let's stick with `smart_text()` to keep this code as it was before.

File: django/contrib/admin/helpers.py (hunk 5 above)
Comment (sasha0) [inline django/contrib/admin/helpers.py]: Ok done.

File: tests/admin_views/tests.py (hunk 3 above)
Comment (timgraham) [inline tests/admin_views/tests.py]: I think that's fine.

File: tests/admin_views/tests.py [hunk 6]
```diff
@@ -4688,6 +4688,24 @@ def test_readonly_backwards_ref(self):
...
+        pd = PlotDetails.objects.create(details='Brand New Plot', plot=pl)
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, 'Brand New Plot')
+
+        # Making sure functionality is not broken when OneToOneField is null.
+        pd.plot = None
+        pd.save()
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        self.assertEqual(response.status_code, 200)
```
Comment (timgraham) [inline tests/admin_views/tests.py]: Can we test for the text here or is it an empty string??

File: docs/releases/1.8.9.txt [hunk 7]
```diff
@@ -26,3 +26,6 @@ Bugfixes
 
 * Fixed a crash when using an ``__in`` lookup inside a ``Case`` expression
   (:ticket:`26071`).
+
+* Fixed a regression that caused exception for backward reference of a OneToOneField
```
Comment (timgraham) [inline docs/releases/1.8.9.txt]: I guess you didn't see my suggestion for the message:
```
* Fixed a crash when using a reverse ``OneToOneField`` in
  ``ModelAdmin.readonly_fields`` (:ticket:`26060`).
```

File: tests/admin_views/tests.py (hunk 6 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Empty. We could check if field is actually there by `self.assertContains(response, 'id_plotdetails')`. Not very reliable, though.

File: docs/releases/1.8.9.txt (hunk 7 above)
Comment (sasha0) [inline docs/releases/1.8.9.txt]: Sorry, missed it. Fixed now.

File: tests/admin_views/tests.py (hunk 3 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Ok, removed it.

File: tests/admin_views/tests.py (hunk 6 above)
Comment (timgraham) [inline tests/admin_views/tests.py]: We could also check `response.context`. Something like (needs to be cleaned up a bit):
```
>>> field_line = [field_line for field_line in [fieldset for fieldset in response.context['adminform']][0]][-1]
>>> for f in field_line: print(f.contents())
Brand New Plot
```
This is easier to debug when it fails than `assertContains`.

File: tests/admin_views/tests.py [hunk 8]
```diff
@@ -4688,6 +4688,23 @@ def test_readonly_backwards_ref(self):
...
+    def test_readonly_onetoone_backwards_ref(self):
+        """
+        Regression test for OneToOneField backward reference, when related field is read-only.
+        """
+        v1 = Villain.objects.create(name='Adam')
+        pl = Plot.objects.create(name='Test Plot', team_leader=v1, contact=v1)
+        pd = PlotDetails.objects.create(details='Brand New Plot', plot=pl)
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, 'Brand New Plot')
```
Comment (timgraham) [inline tests/admin_views/tests.py]: Actually `assertContains` checks the `status_code` too, so the other assertion is redundant.

File: django/contrib/admin/helpers.py (hunk 5 above)
Comment (timgraham) [inline django/contrib/admin/helpers.py]: The test doesn't fail if `value is not None` is removed so maybe it's not needed.

Comment (sasha0) [inline django/contrib/admin/helpers.py]: Agree.

File: tests/admin_views/tests.py (hunk 6 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Ok, reworked checks. Please take a look.

File: django/contrib/admin/helpers.py (hunk 5 above)
Comment (timgraham) [inline django/contrib/admin/helpers.py]: Sorry, what I meant was that the test for the `fk=None` case is useful but `value is not None` seems unneeded.

File: tests/admin_views/tests.py [hunk 9]
```diff
@@ -4688,6 +4688,21 @@ def test_readonly_backwards_ref(self):
...
 
+    def test_readonly_onetoone_backwards_ref(self):
+        """
+        Regression test for OneToOneField backward reference, when related field is read-only.
+        """
+        v1 = Villain.objects.create(name='Adam')
+        pl = Plot.objects.create(name='Test Plot', team_leader=v1, contact=v1)
+        PlotDetails.objects.create(details='Brand New Plot', plot=pl)
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        fieldset = list(response.context['adminform'])[0]
```
Comment (timgraham) [inline tests/admin_views/tests.py]: It's ugly either way, but this way might make it slightly easier to debug if the test fails at some later point.
```python
plot_details_field = list(list(list(response.context['adminform'])[0])[3])[0]
self.assertEqual(plot_details_field.field['name'], 'plotdetails')
self.assertTrue(plot_details_field.is_readonly)
self.assertEqual(plot_details_field.contents(), 'Brand New Plot')
```
I guess it we used the pattern more widely, it might be worth some helper functions to make it easy to extract fields without using magic numbers in the indexing.

File: docs/releases/1.8.9.txt [hunk 10]
```diff
@@ -26,3 +26,7 @@ Bugfixes
 
 * Fixed a crash when using an ``__in`` lookup inside a ``Case`` expression
   (:ticket:`26071`).
+
+* Fixed a crash when using a reverse ``OneToOneField`` in
+  ``ModelAdmin.readonly_fields`` (:ticket:`26060`).
+
```
Comment (timgraham) [inline docs/releases/1.8.9.txt]: chop blank line

File: django/contrib/admin/helpers.py (hunk 5 above)
Comment (sasha0) [inline django/contrib/admin/helpers.py]: Ah ok, no problem, I'l revert.

File: tests/admin_views/tests.py (hunk 9 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: > I guess it we used the pattern more widely, it might be worth some helper functions to make it easy to extract fields without using magic numbers in the indexing.

You mean, it'd be useful to move field extraction code to `django.test.testcases.SimpleTestCase` method?

Comment (timgraham) [inline tests/admin_views/tests.py]: No, it would be admin specific so it doesn't belong there. Just a private API mixin for the Django tests is that I was thinking. It might live in this file, for example.

Comment (sasha0) [inline tests/admin_views/tests.py]: Ok, implemented mixin - please take a look.

**Comment by sasha0** [comment]
@timgraham already updated PR - please take a look.

File: tests/admin_views/tests.py [hunk 11]
```diff
@@ -4688,6 +4718,28 @@ def test_readonly_backwards_ref(self):
...
+    def test_readonly_onetoone_backwards_ref(self):
+        """
+        Regression test for OneToOneField backward reference, when related field is read-only.
+        """
+        v1 = Villain.objects.create(name='Adam')
+        pl = Plot.objects.create(name='Test Plot', team_leader=v1, contact=v1)
+        pd = PlotDetails.objects.create(details='Brand New Plot', plot=pl)
+
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        plot_details_field = self.get_admin_readonly_field(response.context['adminform'], 'plotdetails')
```
Comment (timgraham) [inline tests/admin_views/tests.py]: I guess `get_admin_readonly_field()` could take `response` instead of `response.context['adminform']`.

File: tests/admin_views/tests.py [hunk 12]
```diff
@@ -67,6 +67,36 @@
 for a staff account. Note that both fields may be case-sensitive."
 
 
+class AdminFieldExtractionMixin(object):
+    """Mixin class for extracting data from iterables of `django.contrib.admin.helpers.AdminForm`."""
```
Comment (timgraham) [inline tests/admin_views/tests.py]: please wrap docstrings at 79 characters

File: docs/releases/1.9.2.txt [hunk 13]
```diff
@@ -38,3 +38,7 @@ Bugfixes
 
 * Fixed a crash when destroying an existing test database on MySQL or
   PostgreSQL (:ticket:`26096`).
+  application (:ticket`25858`).
```
Comment (timgraham) [inline docs/releases/1.9.2.txt]: chopx

File: tests/admin_views/tests.py [hunk 14]
```diff
@@ -4688,6 +4718,28 @@ def test_readonly_backwards_ref(self):
...
+        """
+        Regression test for OneToOneField backward reference, when related field is read-only.
+        """
+        v1 = Villain.objects.create(name='Adam')
+        pl = Plot.objects.create(name='Test Plot', team_leader=v1, contact=v1)
+        pd = PlotDetails.objects.create(details='Brand New Plot', plot=pl)
+
+        response = self.client.get(reverse('admin:admin_views_plotproxy_change', args=(pl.pk,)))
+        plot_details_field = self.get_admin_readonly_field(response.context['adminform'], 'plotdetails')
+        self.assertIsNotNone(plot_details_field)
```
Comment (timgraham) [inline tests/admin_views/tests.py]: I don't think this assertion is necessary -- if it's None we'll get an error in the next assertion which should be just as easy to debug.

File: tests/admin_views/tests.py [hunk 15]
```diff
@@ -67,6 +67,36 @@
...
+            field_lines += list(fs)
+
+        fields = []
+        for field_line in field_lines:
+            fields += list(field_line)
+
+        return fields
+
+    def get_admin_readonly_fields(self, admin_form):
+        """Filtering AdminField list to readonly fields."""
```
Comment (timgraham) [inline tests/admin_views/tests.py]: Return the readonly fields for a given AdminForm.
(following verb style of PEP 0257) -- similar for other docstrings

File: tests/admin_views/tests.py [hunk 16]
```diff
@@ -67,6 +67,36 @@
...
 
+class AdminFieldExtractionMixin(object):
+    """Mixin class for extracting data from iterables of `django.contrib.admin.helpers.AdminForm`."""
+
+    def get_admin_form_fields(self, admin_form):
+        """Extracting Fieldset, FieldLine and AdminField objects from AdminForm class."""
+        fieldsets = list(admin_form)
+
+        field_lines = []
+        for fs in fieldsets:
```
Comment (timgraham) [inline tests/admin_views/tests.py]: fs -> fieldset

File: tests/admin_views/tests.py (hunk 11 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Ok. Passed `admin_form` initially into the method, since it was not intended to extract anything else from response, other than the form.

File: docs/releases/1.9.2.txt (hunk 13 above)
Comment (sasha0) [inline docs/releases/1.9.2.txt]: Fixed. Wrong merge somehow.

File: tests/admin_views/tests.py (hunk 16 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Ok. I prefer abbreviated variable naming, when it does not worsen code readability, like this - https://github.com/django/django/blob/e519aab43a419589e92fe284e4ce2f2e034aec6a/django/core/management/commands/showmigrations.py#L100.

**Comment by timgraham** [comment]
Some wording edits and possible alternate fix: http://dpaste.com/0R1HJNV

**Comment by sasha0** [comment]
@timgraham agree about wording.

What pros of alternative fix? Wouldn't it better to return to previous implementation, which we initially intended to?

**Comment by timgraham** [comment]
I'm not sure if there's a good reason for special casing fields in `contents()` instead of delegating to `display_for_field()` as was done in earlier versions.

**Comment by sasha0** [comment]
> I'm not sure if there's a good reason for special casing fields in contents() instead of delegating to display_for_field() as was done in earlier versions.

Hm, yes, it's more logical this way. Will follow your suggestion.

Last question. Instead of this:
```python
-            if field.field.get('name', '') == field_name:
+            if field.field.get('name') == field_name:
                 return field
```
Why not just `field.field['name']` then ? Since `name` key would always be in the `field.field` dict?

**Comment by timgraham** [comment]
Fine with me -- wasn't sure if there was a reason you used `.get()`.

**Comment by sasha0** [comment]
> Fine with me -- wasn't sure if there was a reason you used .get().

To be frankly, just reinsure from `KeyError`.

Ok, I've updated PR considering your suggested edits.

**Comment by timgraham** [comment]
merged in 9a33d3d76497d9e198de942ee1236c452231262f, thanks!
